### PR TITLE
Regenerate client certificate assertions after they expire

### LINKED ISSUE #312: Confidential client doesn't regenerate certificate assertions
There are a couple causes. The operative one today is that `Client.Credential()` only calls `Credential.JWT()` until that method succeeds. Once `Credential` has generated an assertion, `Client` uses that assertion ever after:
https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/e1614e74a7270690a127aa40ddb60f70063d59de/apps/internal/oauth/oauth.go#L104-L109

If it were called again, `Credential.JWT()` would check whether the assertion has expired. However, it simply returns expired assertions:
https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/e1614e74a7270690a127aa40ddb60f70063d59de/apps/internal/oauth/ops/accesstokens/accesstokens.go#L109-L111

### PR DESCRIPTION
**Pull request by chlowell**
This makes `accesstokens.Credential` fully responsible for providing assertions and ensures `Credential.JWT()` generates new assertions as needed from client certificates. This fixes #312 but maintains the current behavior for user-provided assertions--`JWT()` returns these after they expire. I left that as is because `JWT()` can't generate replacements for user-provided assertions and there's no API for applications to do so (#292).

### CODE REVIEW

**Review by jhendrixMSFT — APPROVED** [review]
LGTM but I'd like an ack from an MSAL expert.

**Review by keegan-caruso — APPROVED** [review]

**Comment by element-of-surprise** [comment]
Hey guys, 

Don't wait for me on this one.  If Keegan + Henrik are good, I'm good.  I'm currently getting ready to transit the country and finalizing my book before it releases, so I'm a little out of sorts for extra reviews ATM.

File: apps/internal/oauth/ops/accesstokens/accesstokens.go [hunk 1]
```diff
@@ -106,14 +109,18 @@ func (c *Credential) JWT(authParams authority.AuthParams) (string, error) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	if c.Expires.Before(time.Now()) && c.Assertion != "" {
+	if c.Expires.After(time.Now()) {
```
Comment (bgavrilMS) [inline apps/internal/oauth/ops/accesstokens/accesstokens.go:112]: Can we stop caching assertions completely?  We should rely on token caching not on assertion caching. 

An assertion is not tied just to a certificate, to re-create it you need the `client_id` and the `tenant_id` (for normal AAD assertions). There will be other assertions like ManagedIdentity assertions, which are different.

Comment (bgavrilMS) [reply apps/internal/oauth/ops/accesstokens/accesstokens.go:112]: (we've had quite a few bugs because we were caching assertions)

Comment (chlowell) [reply apps/internal/oauth/ops/accesstokens/accesstokens.go:112]: > Can we stop caching assertions completely? We should rely on token caching not on assertion caching.
> 

Sure, can it wait for a future PR? I'm making a tactical fix here to get certificate authentication working without breaking anything else. Removing or unexporting the `Expires` or `Assertion` fields entails rewriting unrelated tests and maybe some broad refactoring. I'm uncertain of the side effects but don't think they would be trivial.

**Review by bgavrilMS — APPROVED** [review]

**Pull request merged**